<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A short introduction to *MSnbase* development • MSnbase</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="A short introduction to *MSnbase* development">
<meta property="og:description" content="MSnbase">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">MSnbase</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.17.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/v01-MSnbase-demo.html">MSnbase: MS data processing, visualisation and quantification</a>
    </li>
    <li>
      <a href="../articles/v02-MSnbase-io.html">MSnbase IO capabilities</a>
    </li>
    <li>
      <a href="../articles/v03-MSnbase-centroiding.html">MSnbase: centroiding of profile-mode MS data</a>
    </li>
    <li>
      <a href="../articles/v04-benchmarking.html">MSnbase benchmarking</a>
    </li>
    <li>
      <a href="../articles/v05-MSnbase-development.html">A short introduction to *MSnbase* development</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/lgatto/MSnbase/">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="v05-MSnbase-development_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>A short introduction to <em>MSnbase</em> development</h1>
                        <h4 class="author">Laurent Gatto</h4>
            <address class="author_afil">
      de Duve Institute, UCLouvain, Belgium<br><h4 class="author">Johannes Rainer</h4>
            <address class="author_afil">
      Institute for Biomedicine, Eurac Research, Bolzano, Italy.<br><h4 class="author">Sebastian Gibb</h4>
            <address class="author_afil">
      Department of Anesthesiology and Intensive Care, University Medicine Greifswald, Germany.<br><small class="dont-index">Source: <a href="https://github.com/lgatto/MSnbase/blob/master/vignettes/v05-MSnbase-development.Rmd"><code>vignettes/v05-MSnbase-development.Rmd</code></a></small>
      <div class="hidden name"><code>v05-MSnbase-development.Rmd</code></div>

    </address>
</address>
</address>
</div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This vignette describes the classes implemented in  package. It is intended as a starting point for developers or users who would like to learn more or further develop/extend mass spectrometry and proteomics data structures.</p>
    </div>
    
<div id="foreword" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#foreword" class="anchor"></a>Foreword</h1>
<p><a href="http://bioconductor.org/packages/MSnbase"><em>MSnbase</em></a> is under active development; current functionality is evolving and new features will be added. This software is free and open-source software. If you use it, please support the project by citing it in publications:</p>
<blockquote>
<p>Gatto L, Lilley KS. MSnbase-an R/Bioconductor package for isobaric tagged mass spectrometry data visualization, processing and quantitation. Bioinformatics. 2012 Jan 15;28(2):288-9. doi: <a href="https://doi.org/10.1093/bioinformatics/btr645">10.1093/bioinformatics/btr645</a>. PMID: <a href="https://www.ncbi.nlm.nih.gov/pubmed/22113085">22113085</a>.</p>
</blockquote>
</div>
<div id="questions-and-bugs" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#questions-and-bugs" class="anchor"></a>Questions and bugs</h1>
<p>For bugs, typos, suggestions or other questions, please file an issue in our tracking system (<a href="https://github.com/lgatto/MSnbase/issues" class="uri">https://github.com/lgatto/MSnbase/issues</a>) providing as much information as possible, a reproducible example and the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo()</a></code>.</p>
<p>If you don’t have a GitHub account or wish to reach a broader audience for general questions about proteomics analysis using R, you may want to use the Bioconductor support site: <a href="https://support.bioconductor.org/" class="uri">https://support.bioconductor.org/</a>.</p>
<p><strong>NB</strong> This document is going to be updated based on current major development plans in <code>MSnbase</code>.</p>
</div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>This document is not a replacement for the individual manual pages, that document the slots of the <em><a href="https://bioconductor.org/packages/3.13/MSnbase">MSnbase</a></em> classes. It is a centralised high-level description of the package design.</p>
<p><em><a href="https://bioconductor.org/packages/3.13/MSnbase">MSnbase</a></em> aims at being compatible with the <em><a href="https://bioconductor.org/packages/3.13/Biobase">Biobase</a></em> infrastructure <span class="citation">(Gentleman et al. 2004)</span>. Many meta data structures that are used in <code>eSet</code> and associated classes are also used here. As such, knowledge of the <em>Biobase development and the new eSet</em> vignette would be beneficial; the vignette can directly be accessed with <code><a href="https://cran.rstudio.com/web/packages/Biobase/vignettes/BiobaseDevelopment.pdf">vignette("BiobaseDevelopment", package="Biobase")</a></code>.</p>
<p>The initial goal is to use the <em><a href="https://bioconductor.org/packages/3.13/MSnbase">MSnbase</a></em> infrastructure for MS2 labelled (iTRAQ <span class="citation">(Ross et al. 2004)</span> and TMT <span class="citation">(Thompson et al. 2003)</span>) and label-free (spectral counting, index and abundance) quantitation - see the documentation for the <code>quantify</code> function for details. The infrastructure is currently extended to support a wider range of technologies, including metabolomics.</p>
</div>
<div id="coding-style" class="section level1">
<h1 class="hasAnchor">
<a href="#coding-style" class="anchor"></a>Coding style</h1>
<p><code>MSnbase</code> follows the <a href="https://bioconductor.org/developers/how-to/coding-style/">Bioconductor style guide</a>. In particular</p>
<ul>
<li>Do not use <code>.</code> when naming symbols.</li>
<li>A leading <code>.</code> can be used for hidden/local functions or variables.</li>
<li>Snake case should be restricted to internal functions. For consistency, we favour camel case for public functions.</li>
<li>Class names should start with a capital and each class should posses a constructor with identical name. Running the constructor without any input should produce a valid empty object.</li>
<li>Use <code>##</code> to start full-line comments.</li>
<li>For roxygen headers <code>#'</code> is preferred, although <code>##'</code> is tolerated.</li>
<li>Use spaces between <code><a href="https://rdrr.io/r/base/assignOps.html">=</a></code> in function arguments or class definition: <code><a href="https://rdrr.io/pkg/pryr/man/f.html">f(a = 1, b = 2)</a></code>.</li>
<li>Always use a space after a comma: <code>a, b, c</code>.</li>
<li>Always use spaces around binary operators: <code>a + b</code>.</li>
<li>Lines should be kept shorter than 80 characters. For example the following code isn’t accepted</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># no wrap at 80</span>
<span class="va">someVeryLongVariableName</span> <span class="op">&lt;-</span> <span class="fu">someVeryLongFunctionName</span><span class="op">(</span>withSomeEvenLongerFunctionArgumentA <span class="op">=</span> <span class="fl">1</span>, withSomeEvenLongerFunctionArgumentB <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
<p>and should be wrapped as shown below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># alternative 1</span>
<span class="va">someVeryLongVariableName</span> <span class="op">&lt;-</span>
    <span class="fu">someVeryLongFunctionName</span><span class="op">(</span>withSomeEvenLongerFunctionArgumentA <span class="op">=</span> <span class="fl">1</span>,
                             withSomeEvenLongerFunctionArgumentB <span class="op">=</span> <span class="fl">2</span><span class="op">)</span>

<span class="co"># alternative 2</span>
<span class="va">someVeryLongVariableName</span> <span class="op">&lt;-</span> <span class="fu">someVeryLongFunctionName</span><span class="op">(</span>
    withSomeEvenLongerFunctionArgumentA <span class="op">=</span> <span class="fl">1</span>,
    withSomeEvenLongerFunctionArgumentB <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></code></pre></div>
</div>
<div id="msnbase-classes" class="section level1">
<h1 class="hasAnchor">
<a href="#msnbase-classes" class="anchor"></a><em><a href="https://bioconductor.org/packages/3.13/MSnbase">MSnbase</a></em> classes</h1>
<p>All classes have a <code>.__classVersion__</code> slot, of class <code>Versioned</code> from the <em><a href="https://bioconductor.org/packages/3.13/Biobase">Biobase</a></em> package. This slot documents the class version for any instance to be used for debugging and object update purposes. Any change in a class implementation should trigger a version change.</p>
<div id="pset-a-virtual-class-for-raw-mass-spectrometry-data-and-meta-data" class="section level2">
<h2 class="hasAnchor">
<a href="#pset-a-virtual-class-for-raw-mass-spectrometry-data-and-meta-data" class="anchor"></a><code>pSet</code>: a virtual class for raw mass spectrometry data and meta data</h2>
<p>This virtual class is the main container for mass spectrometry data, i.e spectra, and meta data. It is based on the <code>eSet</code> implementation for genomic data. The main difference with <code>eSet</code> is that the <code>assayData</code> slot is an environment containing any number of <code>Spectrum</code> instances (see the <a href="#Spectrum"><code>Spectrum</code> section</a>).</p>
<p>One new slot is introduced, namely <code>processingData</code>, that contains one <code>MSnProcess</code> instance (see the <a href="#MSnProcess"><code>MSnProcess</code> section</a>). and the <code>experimentData</code> slot is now expected to contain <code>MIAPE</code> data. The <code>annotation</code> slot has not been implemented, as no prior feature annotation is known in shotgun proteomics.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"pSet"</span><span class="op">)</span></code></pre></div>
<pre><code>## Virtual Class "pSet" [package "MSnbase"]
## 
## Slots:
##                                                                
## Name:           assayData          phenoData        featureData
## Class:        environment AnnotatedDataFrame AnnotatedDataFrame
##                                                                
## Name:      experimentData       protocolData     processingData
## Class:              MIAxE AnnotatedDataFrame         MSnProcess
##                                             
## Name:              .cache  .__classVersion__
## Class:        environment           Versions
## 
## Extends: "Versioned"
## 
## Known Subclasses: 
## Class "MSnExp", directly
## Class "OnDiskMSnExp", by class "MSnExp", distance 2, with explicit coerce</code></pre>
</div>
<div id="msnexp-a-class-for-ms-experiments" class="section level2">
<h2 class="hasAnchor">
<a href="#msnexp-a-class-for-ms-experiments" class="anchor"></a><code>MSnExp</code>: a class for MS experiments</h2>
<p><code>MSnExp</code> extends <code>pSet</code> to store MS experiments. It does not add any new slots to <code>pSet</code>. Accessors and setters are all inherited from <code>pSet</code> and new ones should be implemented for <code>pSet</code>. Methods that manipulate actual data in experiments are implemented for <code>MSnExp</code> objects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"MSnExp"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "MSnExp" [package "MSnbase"]
## 
## Slots:
##                                                                
## Name:           assayData          phenoData        featureData
## Class:        environment AnnotatedDataFrame AnnotatedDataFrame
##                                                                
## Name:      experimentData       protocolData     processingData
## Class:              MIAxE AnnotatedDataFrame         MSnProcess
##                                             
## Name:              .cache  .__classVersion__
## Class:        environment           Versions
## 
## Extends: 
## Class "pSet", directly
## Class "Versioned", by class "pSet", distance 2
## 
## Known Subclasses: 
## Class "OnDiskMSnExp", directly, with explicit coerce</code></pre>
</div>
<div id="ondiskmsnexp-a-on-disk-implementation-of-the-msnexp-class" class="section level2">
<h2 class="hasAnchor">
<a href="#ondiskmsnexp-a-on-disk-implementation-of-the-msnexp-class" class="anchor"></a><code>OnDiskMSnExp</code>: a on-disk implementation of the <code>MSnExp</code> class</h2>
<p>The <code>OnDiskMSnExp</code> class extends <code>MSnExp</code> and inherits all of its functionality but is aimed to use as little memory as possible based on a balance between memory demand and performance. Most of the spectrum-specific data, like retention time, polarity, total ion current are stored within the object’s <code>featureData</code> slot. The actual M/Z and intensity values from the individual spectra are, in contrast to <code>MSnExp</code> objects, not kept in memory (in the <code>assayData</code> slot), but are fetched from the original files on-demand. Because mzML files are indexed, using the <em><a href="https://bioconductor.org/packages/3.13/mzR">mzR</a></em> package to read the relevant spectrum data is fast and only moderately slower than for in-memory <code>MSnExp</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>.</p>
<p>To keep track of data manipulation steps that are applied to spectrum data (such as performed by methods <code>removePeaks</code> or <code>clean</code>) a <em>lazy execution</em> framework was implemented. Methods that manipulate or subset a spectrum’s M/Z or intensity values can not be applied directly to a <code>OnDiskMSnExp</code> object, since the relevant data is not kept in memory. Thus, any call to a processing method that changes or subset M/Z or intensity values are added as <code>ProcessingStep</code> items to the object’s <code>spectraProcessingQueue</code>. When the spectrum data is then queried from an <code>OnDiskMSnExp</code>, the spectra are read in from the file and all these processing steps are applied on-the-fly to the spectrum data before being returned to the user.</p>
<p>The operations involving extracting or manipulating spectrum data are applied on a per-file basis, which enables parallel processing. Thus, all corresponding method implementations for <code>OnDiskMSnExp</code> objects have an argument <code>BPPARAM</code> and users can set a <code>PARALLEL_THRESH</code> option flag<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> that enables to define how and when parallel processing should be performed (using the <em><a href="https://bioconductor.org/packages/3.13/BiocParallel">BiocParallel</a></em> package).</p>
<p>Note that all data manipulations that are not applied to M/Z or intensity values of a spectrum (e.g. sub-setting by retention time etc) are very fast as they operate directly to the object’s <code>featureData</code> slot.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"OnDiskMSnExp"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "OnDiskMSnExp" [package "MSnbase"]
## 
## Slots:
##                                                                            
## Name:  spectraProcessingQueue                backend              assayData
## Class:                   list              character            environment
##                                                                            
## Name:               phenoData            featureData         experimentData
## Class:     AnnotatedDataFrame     AnnotatedDataFrame                  MIAxE
##                                                                            
## Name:            protocolData         processingData                 .cache
## Class:     AnnotatedDataFrame             MSnProcess            environment
##                              
## Name:       .__classVersion__
## Class:               Versions
## 
## Extends: 
## Class "MSnExp", directly
## Class "pSet", by class "MSnExp", distance 2
## Class "Versioned", by class "MSnExp", distance 3</code></pre>
<p>The distinction between <code>MSnExp</code> and <code>OnDiskMSnExp</code> is often not explicitly stated as it should not matter, from a user’s perspective, which data structure they are working with, as both behave in equivalent ways. Often, they are referred to as <em>in-memory</em> and <em>on-disk</em> <code>MSnExp</code> implementations.</p>
</div>
<div id="msnset-a-class-for-quantitative-proteomics-data" class="section level2">
<h2 class="hasAnchor">
<a href="#msnset-a-class-for-quantitative-proteomics-data" class="anchor"></a><code>MSnSet</code>: a class for quantitative proteomics data</h2>
<p>This class stores quantitation data and meta data after running <code>quantify</code> on an <code>MSnExp</code> object or by creating an <code>MSnSet</code> instance from an external file, as described in the <em>MSnbase-io</em> vignette and in <code><a href="../reference/readMSnSet.html">?readMSnSet</a></code>, <code>readMzTabData</code>, etc. The quantitative data is in form of a <em>n</em> by <em>p</em> matrix, where <em>n</em> is the number of features/spectra originally in the <code>MSnExp</code> used as parameter in <code>quantify</code> and <em>p</em> is the number of reporter ions. If read from an external file, <em>n</em> corresponds to the number of features (protein groups, proteins, peptides, spectra) in the file and <span class="math inline">\(p\)</span> is the number of columns with quantitative data (samples) in the file.</p>
<p>This prompted to keep a similar implementation as the <code>ExpressionSet</code> class, while adding the proteomics-specific annotation slot introduced in the <code>pSet</code> class, namely <code>processingData</code> for objects of class <code>MSnProcess</code>.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"MSnSet"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "MSnSet" [package "MSnbase"]
## 
## Slots:
##                                                                
## Name:      experimentData     processingData               qual
## Class:              MIAPE         MSnProcess         data.frame
##                                                                
## Name:           assayData          phenoData        featureData
## Class:          AssayData AnnotatedDataFrame AnnotatedDataFrame
##                                                                
## Name:          annotation       protocolData  .__classVersion__
## Class:          character AnnotatedDataFrame           Versions
## 
## Extends: 
## Class "eSet", directly
## Class "VersionedBiobase", by class "eSet", distance 2
## Class "Versioned", by class "eSet", distance 3</code></pre>
<p>The <code>MSnSet</code> class extends the virtual <code>eSet</code> class to provide compatibility for <code>ExpressionSet</code>-like behaviour. The experiment meta-data in <code>experimentData</code> is also of class <code>MIAPE</code> . The <code>annotation</code> slot, inherited from <code>eSet</code> is not used. As a result, it is easy to convert <code>ExpressionSet</code> data from/to <code>MSnSet</code> objects with the coersion method <code>as</code>.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">msnset</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "MSnSet"
## attr(,"package")
## [1] "MSnbase"</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="fu"><a href="../reference/as.html">as</a></span><span class="op">(</span><span class="va">msnset</span>, <span class="st">"ExpressionSet"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "ExpressionSet"
## attr(,"package")
## [1] "Biobase"</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "ExpressionSet"
## attr(,"package")
## [1] "Biobase"</code></pre>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/class.html">class</a></span><span class="op">(</span><span class="fu"><a href="../reference/as.html">as</a></span><span class="op">(</span><span class="va">sample.ExpressionSet</span>, <span class="st">"MSnSet"</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "MSnSet"
## attr(,"package")
## [1] "MSnbase"</code></pre>
</div>
<div id="MSnProcess" class="section level2">
<h2 class="hasAnchor">
<a href="#MSnProcess" class="anchor"></a><code>MSnProcess</code>: a class for logging processing meta data</h2>
<p>This class aims at recording specific manipulations applied to <code>MSnExp</code> or <code>MSnSet</code> instances. The <code>processing</code> slot is a <code>character</code> vector that describes major processing. Most other slots are of class <code>logical</code> that indicate whether the data has been centroided, smoothed,  although many of the functionality is not implemented yet. Any new processing that is implemented should be documented and logged here.</p>
<p>It also documents the raw data file from which the data originates (<code>files</code> slot) and the <em><a href="https://bioconductor.org/packages/3.13/MSnbase">MSnbase</a></em> version that was in use when the <code>MSnProcess</code> instance, and hence the <code>MSnExp</code>/<code>MSnSet</code> objects, were originally created.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"MSnProcess"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "MSnProcess" [package "MSnbase"]
## 
## Slots:
##                                                                               
## Name:              files        processing            merged           cleaned
## Class:         character         character           logical           logical
##                                                                               
## Name:       removedPeaks          smoothed           trimmed        normalised
## Class:         character           logical           numeric           logical
##                                           
## Name:     MSnbaseVersion .__classVersion__
## Class:         character          Versions
## 
## Extends: "Versioned"</code></pre>
</div>
<div id="miape-minimum-information-about-a-proteomics-experiment" class="section level2">
<h2 class="hasAnchor">
<a href="#miape-minimum-information-about-a-proteomics-experiment" class="anchor"></a><code>MIAPE</code>: Minimum Information About a Proteomics Experiment</h2>
<p>The Minimum Information About a Proteomics Experiment <span class="citation">(Taylor et al. 2007, 2008)</span> <code>MIAPE</code> class describes the experiment, including contact details, information about the mass spectrometer and control and analysis software.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"MIAPE"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "MIAPE" [package "MSnbase"]
## 
## Slots:
##                                                         
## Name:                     title                      url
## Class:                character                character
##                                                         
## Name:                  abstract                pubMedIds
## Class:                character                character
##                                                         
## Name:                   samples            preprocessing
## Class:                     list                     list
##                                                         
## Name:                     other                dateStamp
## Class:                     list                character
##                                                         
## Name:                      name                      lab
## Class:                character                character
##                                                         
## Name:                   contact                    email
## Class:                character                character
##                                                         
## Name:           instrumentModel   instrumentManufacturer
## Class:                character                character
##                                                         
## Name:  instrumentCustomisations             softwareName
## Class:                character                character
##                                                         
## Name:           softwareVersion        switchingCriteria
## Class:                character                character
##                                                         
## Name:            isolationWidth            parameterFile
## Class:                  numeric                character
##                                                         
## Name:                 ionSource         ionSourceDetails
## Class:                character                character
##                                                         
## Name:                  analyser          analyserDetails
## Class:                character                character
##                                                         
## Name:              collisionGas        collisionPressure
## Class:                character                  numeric
##                                                         
## Name:           collisionEnergy             detectorType
## Class:                character                character
##                                                         
## Name:       detectorSensitivity        .__classVersion__
## Class:                character                 Versions
## 
## Extends: 
## Class "MIAxE", directly
## Class "Versioned", by class "MIAxE", distance 2</code></pre>
</div>
<div id="Spectum" class="section level2">
<h2 class="hasAnchor">
<a href="#Spectum" class="anchor"></a><code>Spectrum</code> <em>et al.</em>: classes for MS spectra</h2>
<p><code>Spectrum</code> is a virtual class that defines common attributes to all types of spectra. MS1 and MS2 specific attributes are defined in the <code>Spectrum1</code> and <code>Spectrum2</code> classes, that directly extend <code>Spectrum</code>.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"Spectrum"</span><span class="op">)</span></code></pre></div>
<pre><code>## Virtual Class "Spectrum" [package "MSnbase"]
## 
## Slots:
##                                                                               
## Name:            msLevel        peaksCount                rt    acquisitionNum
## Class:           integer           integer           numeric           integer
##                                                                               
## Name:          scanIndex               tic                mz         intensity
## Class:           integer           numeric           numeric           numeric
##                                                                               
## Name:           fromFile        centroided          smoothed          polarity
## Class:           integer           logical           logical           integer
##                         
## Name:  .__classVersion__
## Class:          Versions
## 
## Extends: "Versioned"
## 
## Known Subclasses: "Spectrum2", "Spectrum1"</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"Spectrum1"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "Spectrum1" [package "MSnbase"]
## 
## Slots:
##                                                                               
## Name:            msLevel        peaksCount                rt    acquisitionNum
## Class:           integer           integer           numeric           integer
##                                                                               
## Name:          scanIndex               tic                mz         intensity
## Class:           integer           numeric           numeric           numeric
##                                                                               
## Name:           fromFile        centroided          smoothed          polarity
## Class:           integer           logical           logical           integer
##                         
## Name:  .__classVersion__
## Class:          Versions
## 
## Extends: 
## Class "Spectrum", directly
## Class "Versioned", by class "Spectrum", distance 2</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"Spectrum2"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "Spectrum2" [package "MSnbase"]
## 
## Slots:
##                                                                
## Name:              merged        precScanNum        precursorMz
## Class:            numeric            integer            numeric
##                                                                
## Name:  precursorIntensity    precursorCharge    collisionEnergy
## Class:            numeric            integer            numeric
##                                                                
## Name:             msLevel         peaksCount                 rt
## Class:            integer            integer            numeric
##                                                                
## Name:      acquisitionNum          scanIndex                tic
## Class:            integer            integer            numeric
##                                                                
## Name:                  mz          intensity           fromFile
## Class:            numeric            numeric            integer
##                                                                
## Name:          centroided           smoothed           polarity
## Class:            logical            logical            integer
##                          
## Name:   .__classVersion__
## Class:           Versions
## 
## Extends: 
## Class "Spectrum", directly
## Class "Versioned", by class "Spectrum", distance 2</code></pre>
</div>
<div id="reporterions-a-class-for-isobaric-tags" class="section level2">
<h2 class="hasAnchor">
<a href="#reporterions-a-class-for-isobaric-tags" class="anchor"></a><code>ReporterIons</code>: a class for isobaric tags</h2>
<p>The iTRAQ and TMT (or any other peak of interest) are implemented <code>ReporterIons</code> instances, that essentially defines an expected MZ position for the peak and a width around this value as well a names for the reporters.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"ReporterIons"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "ReporterIons" [package "MSnbase"]
## 
## Slots:
##                                                                               
## Name:               name     reporterNames       description                mz
## Class:         character         character         character           numeric
##                                                             
## Name:                col             width .__classVersion__
## Class:         character           numeric          Versions
## 
## Extends: "Versioned"</code></pre>
</div>
<div id="chromatogram-and-mchromatograms-classes-to-handle-chromatographic-data" class="section level2">
<h2 class="hasAnchor">
<a href="#chromatogram-and-mchromatograms-classes-to-handle-chromatographic-data" class="anchor"></a><code>Chromatogram</code> and <code>MChromatograms</code>: classes to handle chromatographic data</h2>
<p>The <code>Chromatogram</code> class represents chromatographic MS data, i.e. retention time and intensity duplets for one file/sample. The <code>MChromatograms</code> class (Matrix of Chromatograms) allows to arrange multiple <code>Chromatogram</code> instances in a two-dimensional grid, with columns supposed to represent different samples and rows two-dimensional areas in the plane spanned by the m/z and retention time dimensions from which the intensities are extracted (e.g. an extracted ion chromatogram for a specific ion). The <code>MChromatograms</code> class extends the base <code>matrix</code> class. <code>MChromatograms</code> objects can be extracted from an <code>MSnExp</code> or <code>OnDiskMSnExp</code> object using the <code>chromatogram</code> method.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"Chromatogram"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "Chromatogram" [package "MSnbase"]
## 
## Slots:
##                                                                               
## Name:              rtime         intensity                mz          filterMz
## Class:           numeric           numeric           numeric           numeric
##                                                                               
## Name:        precursorMz         productMz          fromFile    aggregationFun
## Class:           numeric           numeric           integer         character
##                                           
## Name:            msLevel .__classVersion__
## Class:           integer          Versions
## 
## Extends: "Versioned"</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"MChromatograms"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "MChromatograms" [package "MSnbase"]
## 
## Slots:
##                                                                
## Name:               .Data          phenoData        featureData
## Class:             matrix AnnotatedDataFrame AnnotatedDataFrame
## 
## Extends: 
## Class "matrix", from data part
## Class "array", by class "matrix", distance 2
## Class "structure", by class "matrix", distance 3
## Class "vector", by class "matrix", distance 4, with explicit coerce
## Class "vector_OR_Vector", by class "matrix", distance 5, with explicit coerce
## Class "vector_OR_factor", by class "matrix", distance 5, with explicit coerce</code></pre>
</div>
<div id="other-classes" class="section level2">
<h2 class="hasAnchor">
<a href="#other-classes" class="anchor"></a>Other classes</h2>
<div id="lists-of-msnset-instances" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#lists-of-msnset-instances" class="anchor"></a>Lists of <code>MSnSet</code> instances</h3>
<p>When several <code>MSnSet</code> instances are related to each other and should be stored together as different objects, they can be grouped as a list into and <code>MSnSetList</code> object. In addition to the actual <code>list</code> slot, this class also has basic logging functionality and enables iteration over the <code>MSnSet</code> instances using a dedicated <code>lapply</code> methods.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">getClass</span><span class="op">(</span><span class="st">"MSnSetList"</span><span class="op">)</span></code></pre></div>
<pre><code>## Class "MSnSetList" [package "MSnbase"]
## 
## Slots:
##                                                                               
## Name:                  x               log       featureData .__classVersion__
## Class:              list              list         DataFrame          Versions
## 
## Extends: "Versioned"</code></pre>
</div>
</div>
</div>
<div id="miscellaneous" class="section level1">
<h1 class="hasAnchor">
<a href="#miscellaneous" class="anchor"></a>Miscellaneous</h1>
<div id="unit-tests" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#unit-tests" class="anchor"></a>Unit tests</h4>
<p><em><a href="https://bioconductor.org/packages/3.13/MSnbase">MSnbase</a></em> implements unit tests with the <em><a href="https://CRAN.R-project.org/package=testthat">testthat</a></em> package.</p>
</div>
<div id="processing-methods" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#processing-methods" class="anchor"></a>Processing methods</h4>
<p>Methods that process raw data, i.e. spectra should be implemented for <code>Spectrum</code> objects first and then <code>eapply</code>ed (or similar) to the <code>assayData</code> slot of an <code>MSnExp</code> instance in the specific method.</p>
</div>
</div>
<div id="session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#session-information" class="anchor"></a>Session information</h1>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2020-12-24 r79679)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] MSnbase_2.17.3      ProtGenerics_1.23.5 S4Vectors_0.29.6   
## [4] mzR_2.25.1          Rcpp_1.0.5          Biobase_2.51.0     
## [7] BiocGenerics_0.37.0 BiocStyle_2.19.1   
## 
## loaded via a namespace (and not attached):
##  [1] lattice_0.20-41       foreach_1.5.1         assertthat_0.2.1     
##  [4] rprojroot_2.0.2       digest_0.6.27         R6_2.5.0             
##  [7] plyr_1.8.6            mzID_1.29.0           evaluate_0.14        
## [10] ggplot2_3.3.2         pillar_1.4.7          zlibbioc_1.37.0      
## [13] rlang_0.4.9           preprocessCore_1.53.1 rmarkdown_2.6        
## [16] pkgdown_1.6.1.9000    textshaping_0.2.1     desc_1.2.0           
## [19] BiocParallel_1.25.2   stringr_1.4.0         munsell_0.5.0        
## [22] compiler_4.1.0        xfun_0.19             pkgconfig_2.0.3      
## [25] systemfonts_0.3.2     pcaMethods_1.83.0     htmltools_0.5.0      
## [28] tidyselect_1.1.0      tibble_3.0.4          bookdown_0.21        
## [31] IRanges_2.25.6        codetools_0.2-18      XML_3.99-0.5         
## [34] crayon_1.3.4          dplyr_1.0.2           MASS_7.3-53          
## [37] grid_4.1.0            gtable_0.3.0          lifecycle_0.2.0      
## [40] affy_1.69.0           magrittr_2.0.1        MsCoreUtils_1.3.2    
## [43] scales_1.1.1          ncdf4_1.17            stringi_1.5.3        
## [46] impute_1.65.0         fs_1.5.0              affyio_1.61.0        
## [49] doParallel_1.0.16     limma_3.47.3          ellipsis_0.3.1       
## [52] ragg_0.4.0            generics_0.1.0        vctrs_0.3.6          
## [55] iterators_1.0.13      tools_4.1.0           glue_1.4.2           
## [58] purrr_0.3.4           yaml_2.2.1            colorspace_2.0-0     
## [61] BiocManager_1.30.10   vsn_3.59.0            MALDIquant_1.19.3    
## [64] memoise_1.1.0         knitr_1.30</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-Gentleman2004">
<p>Gentleman, Robert C., Vincent J. Carey, Douglas M. Bates, Ben Bolstad, Marcel Dettling, Sandrine Dudoit, Byron Ellis, et al. 2004. “Bioconductor: Open Software Development for Computational Biology and Bioinformatics.” <em>Genome Biol</em> 5 (10): –80. <a href="https://doi.org/10.1186/gb-2004-5-10-r80">https://doi.org/10.1186/gb-2004-5-10-r80</a>.</p>
</div>
<div id="ref-Ross2004">
<p>Ross, Philip L., Yulin N. Huang, Jason N. Marchese, Brian Williamson, Kenneth Parker, Stephen Hattan, Nikita Khainovski, et al. 2004. “Multiplexed Protein Quantitation in Saccharomyces Cerevisiae Using Amine-Reactive Isobaric Tagging Reagents.” <em>Mol Cell Proteomics</em> 3 (12): 1154–69. <a href="https://doi.org/10.1074/mcp.M400129-MCP200">https://doi.org/10.1074/mcp.M400129-MCP200</a>.</p>
</div>
<div id="ref-Taylor2008">
<p>Taylor, Chris F, Pierre-Alain Binz, Ruedi Aebersold, Michel Affolter, Robert Barkovich, Eric W Deutsch, David M Horn, et al. 2008. “Guidelines for Reporting the Use of Mass Spectrometry in Proteomics.” <em>Nat. Biotechnol.</em> 26 (8): 860–1. <a href="https://doi.org/10.1038/nbt0808-860">https://doi.org/10.1038/nbt0808-860</a>.</p>
</div>
<div id="ref-Taylor2007">
<p>Taylor, Chris F., Norman W. Paton, Kathryn S. Lilley, Pierre-Alain Binz, Randall K. Julian, Andrew R. Jones, Weimin Zhu, et al. 2007. “The Minimum Information About a Proteomics Experiment (Miape).” <em>Nat Biotechnol</em> 25 (8): 887–93. <a href="https://doi.org/10.1038/nbt1329">https://doi.org/10.1038/nbt1329</a>.</p>
</div>
<div id="ref-Thompson2003">
<p>Thompson, Andrew, Jürgen Schäfer, Karsten Kuhn, Stefan Kienle, Josef Schwarz, Günter Schmidt, Thomas Neumann, R Johnstone, A Karim A Mohammed, and Christian Hamon. 2003. “Tandem Mass Tags: A Novel Quantification Strategy for Comparative Analysis of Complex Protein Mixtures by MS/MS.” <em>Anal. Chem.</em> 75 (8): 1895–1904.</p>
</div>
</div>
</div>
<div class="footnotes">
<hr>
<ol>
<li id="fn1"><p>The <em>benchmarking</em> vignette compares data size and operation speed of the two implementations.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>see <code><a href="../reference/MSnbaseOptions.html">?MSnbaseOptions</a></code> for details.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Laurent Gatto, Johannes Rainer, Sebastian Gibb.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.9000.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
